### PR TITLE
Protect file from corruption while saving

### PR DESCRIPTION
**Pull request by bokobza**
This should fix https://github.com/stratisproject/StratisBitcoinFullNode/issues/1230 and hopefully https://github.com/stratisproject/StratisBitcoinFullNode/issues/1231.

I added a call to `File.Replace` when saving the file so that the file never gets in a corrupted state.
This call is atomic in the sense that the file is either replaced successfully or not replaced at all.

Additionally, I added a parameter to the `FileStorage.SaveFile` method where a backup file (.bak) will be saved every time the file is saved. So the backup will have the previous version, and the file will have the new version.

More here: https://msdn.microsoft.com/en-us/library/windows/desktop/hh802690%28v=vs.85%29.aspx

### CODE REVIEW

**Comment by Aprogiena** [comment]
The ultimate solution to this is to use DBreeze, but for now using file replace is probably good enough (although documentation does not really guarantees File.Replace to be atomic). However, you also have this 

```
 // If the file does not exist yet, create it.
+            if (!File.Exists(filePath))
+            {
+                File.WriteAllText(filePath, JsonConvert.SerializeObject(toSave, Formatting.Indented));
```

so you expect File.WriteAllText to be atomic, which it probably is not.

**Comment by fassadlr** [comment]
@bokobza Why don't we just always save a back up file? Or, in what situations would we NOT want to save a back  up?

If we dont always want to save a back up file via the wallet manager then in what instances do we want to?

**Comment by bokobza** [comment]
@Aprogiena The documentation states: 
> The needed atomicity here is that the changes either are completely applied or not applied at all, as an inconsistent state would render the file corrupt. 


> Why don't we just always save a back up file?

@fassadlr oh just because it's a few more operations. Also, when you call a method called SaveFile, you're not expecting a backup to be created for you by default. It's there if you want to use it though.
We could use a backup feature for example if we want to save a backup of a newly created wallet. This way if the wallet gets corrupted, we can just delete it and use the backup.


For File.WriteAllText, true, it's not guaranteed to be atomic. However, here we use it 1. on the creation of the file and 2. to create the temp file (when the file already exists).
In both occasions, if it fails, no big deal. In the first instance, the probability that the system will crash the first time a file is created is very slim, and in the second instance, only the temp file will be corrupted, the main file staying at is it.

I can do 2 things:
1. on creation, create a tmp file and then rename it -> if it fails while creating the tmp file, no big deal.
2. find another way to write the file atomically. I've searched for an out-of-the-box solution but couldn't find any.

In any case I'll also add a lock around the save operation.

**Comment by Aprogiena** [comment]
@bokobza Okay, I'm happy with replace file.

So just change write all text. Of course the probability is low, but that's not a good argument, it will just happen later. Simply avoid any chance of corruption. Since it is not possible to write the file atomically, just use option one as you suggested.

**Comment by bokobza** [comment]
you got it baby.

**Comment by bokobza** [comment]
Works on Mac too.

File: src/Stratis.Bitcoin.Tests/Utilities/FileStorageTest.cs [hunk 1]
```diff
@@ -227,6 +227,103 @@ public void GivenLoadByFileExtensionIsCalled_WhenFilesDontExist_ThenNoObjectsIsR
...
+
+            // Assert
+            Assert.True(isFileExists);
+            Assert.False(isBackupFileExists);
+            TestObject loadedByFileName = fileStorage.LoadByFileName("savedTestObject.json");
+            Assert.Equal("prop1-changed", loadedByFileName.Property1);
+            Assert.Equal("prop2", loadedByFileName.Property2);
+
+        }
+
```
Comment (monsieurleberre) [inline src/Stratis.Bitcoin.Tests/Utilities/FileStorageTest.cs]: Not sure it adds much value, but maybe a separate test to check the default behavior is false, and a test to see what happens when the false is used explicitly ...

File: src/Stratis.Bitcoin/Utilities/FileStorage.cs [hunk 2]
```diff
@@ -29,18 +29,37 @@ public FileStorage(string folderPath)
...
         /// <param name="fileName">Name of the file to be saved.</param>
-        public void SaveToFile(T toSave, string fileName)
+        /// <param name="saveBackupFile">A value indicating whether to save a backup of the file.</param>
+        public void SaveToFile(T toSave, string fileName, bool saveBackupFile = false)
         {
             Guard.NotEmpty(fileName, nameof(fileName));
             Guard.NotNull(toSave, nameof(toSave));
 
             string filePath = Path.Combine(this.FolderPath, fileName);
+            string tempFilePath = $"{filePath}.temp";
```
Comment (monsieurleberre) [inline src/Stratis.Bitcoin/Utilities/FileStorage.cs]: I would get a random temp name using something like System.IO.Path.GetTempFileName(); so you reduce the chances of writing to a file that is already being locked or used in some way ...

**Comment by monsieurleberre** [comment]
My suggestions are only suggestions, I am happy for that to go in with or without them :) 🍃 🦄 🏈

File: src/Stratis.Bitcoin/Utilities/FileStorage.cs (hunk 2 above)
Comment (bokobza) [reply src/Stratis.Bitcoin/Utilities/FileStorage.cs]: That's a great suggestion. I'm keen to keep the temp file in the same folder as the principal so GetTempFileName won't work for us, but I'll find a way to create unique temp file names (probably using ticks) .

**Comment by monsieurleberre** [comment]
bang on 💸

**Pull request merged**